src/gui/guiFormSpecMenu.cpp

```diff
@@ -1679,22 +1679,60 @@ void GUIFormSpecMenu::parseHyperText(parserData *data, const std::string &elemen
 
 void GUIFormSpecMenu::parseLabel(parserData* data, const std::string &element)
 {
-	std::vector<std::string> parts = split(element,';');
+	std::vector<std::string> parts = split(element, ';');
 
 	if ((parts.size() == 2) ||
 		((parts.size() > 2) && (m_formspec_version > FORMSPEC_API_VERSION)))
 	{
-		std::vector<std::string> v_pos = split(parts[0],',');
+		std::vector<std::string> v_pos = split(parts[0], ',');
 		std::string text = parts[1];
 
-		MY_CHECKPOS("label",0);
+		MY_CHECKPOS("label", 0);
 
-		if(!data->explicit_size)
-			warningstream<<"invalid use of label without a size[] element"<<std::endl;
+		if (!data->explicit_size)
+			warningstream << "Invalid use of label without a size[] element." << std::endl;
 
 		std::vector<std::string> lines = split(text, '\n');
+		auto style = getStyleForElement("label", "");
+
+		EGUI_ALIGNMENT horigin = gui::EGUIA_LOWERRIGHT;
+		EGUI_ALIGNMENT vorigin = gui::EGUIA_SCALE; // Used as base
+		std::string str_horigin = trim(style.get(StyleSpec::HORIGIN, ""));
+		std::string str_vorigin = trim(style.get(StyleSpec::VORIGIN, ""));
+
+		if (str_horigin == "left")
+			horigin = gui::EGUIA_UPPERLEFT;
+		else if (str_horigin == "center")
+			horigin = gui::EGUIA_CENTER;
+		else if (str_horigin == "base")
+			horigin = gui::EGUIA_SCALE;
+
+		if (str_vorigin == "top")
+			vorigin = gui::EGUIA_UPPERLEFT;
+		else if (str_vorigin == "center")
+			vorigin = gui::EGUIA_CENTER;
+		else if (str_vorigin == "bottom")
+			vorigin = gui::EGUIA_LOWERRIGHT;
+
+		s32 font_height = font_line_height(m_font);
+
+		std::string str_newline = style.get(StyleSpec::NEWLINE_SPACING, "0.5");
+		float newline_spacing;
+		if (trim(str_newline) == "auto")
+			newline_spacing = (float)font_height;
+		else
+			newline_spacing = stof(str_newline) * (float)imgsize.Y;
```

Comment:
Actually, strangely enough, `stof` and `stoi` don't give errors; if the number is invalid, it returns `0`.  Try inputting an invalid number to anything that takes a number, e.g. any element position.  No element positioner does error checking: https://github.com/minetest/minetest/blob/master/src/gui/guiFormSpecMenu.cpp#L280-L299.  I don't know why it does this, but I'm kind of glad it does.

`stof("110%")` will return `110.0`; it searches up until the first invalid character.

Reply:
```C++
bool ok = sscanf(str_newline, "%f", &newline_spacing);
if (!ok)
	newline_spacing = 0;
newline_spacing += imgsize.Y;
```
This would be the sophisticated approach.

Reply:
I don't see what's the point of this when _every_ other formspec element uses `stoi` and `stof`, relying on the returning `0` behaviour.  If it gets changed, it should be changed for everything.

But that `sscanf` approach is pretty nice, all the same.

Reply:
I found the offenders.  It isn't using `std::stoi` or `std::stof` after all (I knew it seemed strange because these are C++11 which Minetest wasn't officially using until a little while ago, but tons of ancient code uses it) but `atoi` and `atof`:

https://github.com/minetest/minetest/blob/0e88176db8ce9f3fb631feb56bdd29d4ada054f5/src/util/string.h#L342-L391